### ISSUE TITLE
Resultados dos Experimentos - Otimizadores

### ISSUE THREAD

**Issue by liviameinhardt** [liviameinhardt/deep-learning-course-fgv#1]
Poste seus resultados aqui. Lembre de colocar:

- Seu nome e sobrenome;
- Perguntas/Hipóteses que quis responder;
- Imagem dos gráficos gerados;
- Breve comentário explicando observações.

**Comment by LarissaLAfonso** [comment]
- Larissa Lemos Afonso
- Queria entender como o Adam se comportava com diferentes taxas de aprendizado

<img width="1383" height="639" alt="Image" src="https://github.com/user-attachments/assets/edabe8da-a7c1-4efe-91d6-c39c5ef638ca" />

<img width="1356" height="640" alt="Image" src="https://github.com/user-attachments/assets/fd6a0eed-f6bb-4911-8a37-495c28c1910a" />

- Observando as imagens (e os logs), percebi que lr = 0.1 é uma taxa muito grande para a geometria da função de perda desses dados; os resultados, além de mudarem muito mais lentamente, parecem não convergir, oscilando em torno de um ponto. Eu diria que isso provavelmente significa que o gradiente caiu em algum vale (que é único, dado que a perda é quadrática), e que o passo é grande demais para que ele consiga efetivamente convergir para o mínimo.

**Comment by kauanmaf** [comment]
<img width="1990" height="490" alt="Image" src="https://github.com/user-attachments/assets/9de8ac17-3f1f-41bb-b6aa-80bd1d24aec0" />

Na figura, foi testado diferentes valores de learning rate para o otimizador Adam. Observa-se que o maior learning rate no qual a rede ainda consegue aprender de forma estável é 0.01. Para valores acima disso, o modelo não converge bem, o que resulta em baixa acurácia e altas perdas no treino. Além disso, um learning rate muito elevados, especialmente acima de 1, leva a um comportamento instável, piorando e muito a capacidade de aprendizado da rede.

**Comment by Vinasque** [comment]
**Nome:** Vinícius Nascimento
**Pergunta:** Com a mesma learning rate e o mesmo momentum, Nesterov converge mais rápido que Momentum clássico no FashionMNIST?

<img width="1986" height="490" alt="Image" src="https://github.com/user-attachments/assets/40e5c3ae-9584-445d-b9f9-5e7c2c0adeba" />

**Comentário:** o Nesterov apresentou convergência mais rápida que o momentum clássico no FashionMNIST. Para isso, adaptei o código de forma que contasse o tempo em que o algoritmo executou até alcançar os 85% de acurácia. Enquanto o SGD com momentum alcançou 85% de acurácia em cerca de 24,56 segundos, o Nesterov atingiu esse valor em aproximadamente 16,62 segundos, mostrando maior eficiência. Além disso, o Nesterov obteve perdas de treino mais baixas e acurácias finais um pouco superiores em comparação ao momentum clássico. Já o SGD puro, sem momentum, não conseguiu chegar a 85% de acurácia dentro das 10 épocas avaliadas. De certa forma, esse teste reforçou ainda mais a importância de estratégias de aceleração no processo de otimização.

**Comment by GabrielAbad** [comment]
- Gabriel Rodrigues de Deus Abad
- Pergunta/Hipótese que quis responder;
  1. Uma taxa de aprendizado (learning rate) mais alta no otimizador Adagrad leva a uma convergência mais rápida, mas prejudica a capacidade de generalização do modelo?

<img width="1840" height="454" alt="Image" src="https://github.com/user-attachments/assets/f4346d3d-6cf1-4eec-9ebd-9eab21fd7003" />

- Breve comentário explicando observações.
    - Os gráficos indicam que a taxa de aprendizado maior (lr=0.3) converge a perda de treino para zero já na 1ª epoch, porém resulta em uma acurácia no teste significativamente menor e mais instável. Mas, a taxa de aprendizado menor (lr=0.01) também minimiza a perda eficientemente e alcança uma acurácia no teste superior e mais consistente. Isso sugere que um learning rate muito alto pode fazer o otimizador passar de um bom mínimo local, prejudicando o desempenho do modelo em dados não vistos.

**Comment by AlessandraBello** [comment]
- Alessandra Belló Soares
- O que acontece quando a Learning Rate é alta e qual otimizador tem melhor desempenho?
- Imagem dos gráficos gerados:

<img width="1729" height="436" alt="Image" src="https://github.com/user-attachments/assets/add629fa-3dfb-48ae-8ba2-6a2845cbd445" />

- Como podemos ver, neste caso, o Adagrad foi melhor tanto para taxas de aprendizado altas quanto baixas. O que podemos observar é que quando a taxa de aprendizado é maior a acurácia no teste fica muito abaixa, diferente de quando ela é baixa e a acurácia chega a 0.89 com o Adagrad.

**Comment by vitor-n** [comment]
**Vitor do Nascimento**

**Pergunta:** Como é diferentes otimizadores se comportam para um mesmo _Learning Rate_?

<img width="1489" height="490" alt="Image" src="https://github.com/user-attachments/assets/7bcbbc31-8873-4eff-a26f-34b1ecd5402d" />

<img width="300" height="235" alt="Image" src="https://github.com/user-attachments/assets/5c9a2fd5-0463-45d4-abbd-66ba29de41ef" />

**Comentario:** Podemos notar que os otimizadores estão cumprindo sua função de reduzir a perda. No entanto, cada otimizador apresentou uma eficiencia diferente para o problema dado (FashionMNIST). Podemos notar que o RMSProp e o Adam foram os otimizadores mais eficientes dentre os demais. Para muitas aplicações o Adam é a escolha padrão por conta da sua adaptatividade.

**Comment by pedrotokar** [comment]
- Pedro Santos Tokar
- Quis observar se o algorítmo Adam incrementado com o momento do algorítmo do gradiente de Nesterov ([NAdam no pytorch](https://docs.pytorch.org/docs/stable/generated/torch.optim.NAdam.html#torch.optim.NAdam)) supera o Adam padrão na otimização ou se são +- equivalentes

<img width="1989" height="1475" alt="Image" src="https://github.com/user-attachments/assets/3734817f-438a-4f97-b0d6-239a595fd3ed" />

Ao que me parece, o algorítmo NAdam melhora um pouco o treinamento, já que no final dos casos com lr 0,01 e 0,001 ele alcançou uma perda menor no treino e uma acurácia maior nos testes (no caso com lr 0,0001 não pareceu fazer muita diferença, provavelmente pelo learning rate ser muito pequeno e as diferenças então se perderem).
Ao mesmo tempo, ele me pareceu um pouco mais instável do que o Adam, novamente observando os casos com lr 0,01 e 0,001. Em alguns momentos do treinamento ele fica com acurácia menor que o Adam habitual, mas eu precisaria rodar com mais épocas para ver se a acurácia estabilizaria acima da obtida pelo Adam. Levando em conta somente a final, ele parece ser melhor. 
Uma possível melhoria pra essa análise seria olhar se os hiperparâmetros desses otimizadores afetam eles de formas diferentes. Nos testes usei os padrões do pytorch

**Comment by Gustyper** [comment]
**Nome:** Gustavo Bianchi da Silva
**Pergunta:** O Adagrad  pode fazer o learning rate decair muito rapidamente e "congelar" o aprendizado. Como isso se comporta quando temos um learning rate inicial muito alto?

<img width="1811" height="442" alt="Image" src="https://github.com/user-attachments/assets/9ce76ff8-7e59-4e50-8a26-715574605aff" />

**Comentário:** Ao contrário do que eu esperava, um learning rate mais alto no Adagrad congelou o aprendizado mais rápido do que um learning rate baixo. Como as primeiras atualizações foram muito grandes, o acumulado dos gradientes acabou crescendo rapidamente, reduzindo drasticamente o learning rate efetivo desde o início. Como resultado, o modelo passa a aprender mais lentamente, não por estar seguindo uma direção errada, mas devido ao mecanismo adaptativo do otimizador.

**Comment by marimarifr** [comment]
* Mariana Fernandes;
* Quanto o momento pode "segurar" a atualização dos pesos e impactar sua performance?

<img width="882" height="457" alt="Image" src="https://github.com/user-attachments/assets/f426f7df-b38b-41c1-a03d-04244f4280a1" />

<img width="881" height="430" alt="Image" src="https://github.com/user-attachments/assets/8954b042-01ae-4db9-a20a-e51ba31ace06" />
 
* Vemos que nessse treinamento o momento teve um impacto positivo quando seu peso também era maior, a configuração intermediária teve uma performance similar ao otimizador sem momento algum e aumentamos o seu peso e vimos um melhora na acurácia, podemos dizer que com esses resultados mostra que o problema ainda tem espaço a ser explorado e podemos adicionar mais épocas

**Comment by STAN120404** [comment]
Stanislas Michel

Perguntas/Hipóteses que quis responder
Queria comparar Adam vs SGD sem momentum e Adagrad, analisando velocidade de convergência, estabilidade e acurácia final.

<img width="938" height="328" alt="Image" src="https://github.com/user-attachments/assets/b9d75be1-ad41-48a1-b990-40148992989e" />

Adam → converge mais rápido e de forma mais estável, com acurácia final ligeiramente superior.
SGD → precisa de mais épocas ou de um LR maior para se aproximar de Adam; LR adequado é crucial para estabilidade.
Adagrad → desempenho próximo ao Adam, mas LR adaptativo pode gerar instabilidades em alguns batches.
Conclusão → Adam é ótimo para início rápido, mas SGD pode ser competitivo com ajuste correto de LR e momentum.

**Comment by LeonardoAleee** [comment]
• Leonardo Alexandre
• Qual otimizador atinge a maior acurácia em menos épocas, mantendo a mesma configuração de rede e conjunto de dados?• <img width="1319" height="326" alt="Image" src="https://github.com/user-attachments/assets/ea1b0dcc-6de3-456b-8395-fcba7c2f2296" />>
• Observa-se que todos os otimizadores melhoraram a acurácia ao longo das épocas, mas em ritmos diferentes:
   - O Adagrad (lr=0.01) alcançou a melhor acurácia final (~0.892) já na 10ª época, superando os demais otimizadores.
   - O SGD puro foi o mais lento a convergir, enquanto suas variações com Momentum e Nesterov mostraram melhora significativa, mas ainda abaixo dos métodos adaptativos.
   - O RMSProp estabilizou em acurácias intermediárias, sem superar Adagrad e Adam.

**Comment by antoniobrych** [comment]
Pergunta: A aceleração a aceleração observada do Momentum em relação ao SGD simples se aproxima do fator
`1/ (1-alpha)` onde `alpha` é o parâmetro do momentum?

<img width="2388" height="1277" alt="Image" src="https://github.com/user-attachments/assets/9418f5be-af9b-4ae1-853e-59a82e5fb688" />

**Comment by PAULA-123** [comment]
- Paula Eduarda de Lima;
- O que acontece se todos esses modelos estiverem com a mesma taxa de aprendizado?

<img width="1750" height="571" alt="Image" src="https://github.com/user-attachments/assets/2c7f4185-b631-4ce1-a2ff-92e8006755b9" />
<img width="1745" height="571" alt="Image" src="https://github.com/user-attachments/assets/daeb323a-9aab-4766-a566-480a296cf37e" />


Como podemos perceber pelos gráficos nem todo learning rate funciona para diferentes modelos, ou seja, cada modelo tem o seu passo que melhor se adapta e resulta em uma melhor otimização do erro, de acordo com sua arquitetura. Por exemplo, o RMSprop ficou com uma acurácia menor comparada com os outros o que pode mostrar que o passo de 0.01 ainda é pequeno para ele, o que faz não apresentar tanta evolução na acurácia e no erro com poucas interações.

O que é diferente do que acontece quando todos estão com o lr como 0.1. Com esse passo as diferenças são acentuadas muito, com o SGD e o Adagrad se saindo bem melhor.

Isso acontece provavelmente porque cada otimizador implementa uma forma diferente de ajustar os pesos a partir do gradiente, e por isso a escala do learning rate tem efeitos distintos em cada um deles.

**Comment by carvguilherme8** [comment]
- Qual a diferença de comportamento da velocidade de convergência de otimizadores adaptativos em compação ao SGD quando a taxa de aprendizado é baixa.
- Guilherme Carvalho Cunha

<img width="1981" height="486" alt="Image" src="https://github.com/user-attachments/assets/76af4ff9-1a3b-4811-a157-1f4733afea4a" />

Com os resultados obtidos podemos assumir a superioridade de otimizadores adaptativos em relação ao SGD quando a learning rate é baixa. Nesse nosso caso, os modelos adaptativos possivelmente aumentaram o learning rate por conta própria, o que se mostrou efetivo para o conjunto de dados utilizado. Já o SGD se manteve com learning rate constante, observada pela convergência lenta e gradual.

**Comment by gtironi** [comment]
- Gustavo Tironi
- O que é melhor, um learning rate menor ou adicionar momentum no SGD?

<img width="1054" height="254" alt="Image" src="https://github.com/user-attachments/assets/4caf9d70-efcf-482c-bcd6-431cd0657b18" />

No geral, nenhum teve uma performace muito boa. Contudo, é claro que o momentun ajudou muito mais do que a diminuição do learning rate na estabilização da descida. Podemos ver que a diminuição do leaning rate desacelerou muito o aprendizado, enquanto o momentum estabilizou, mas ainda mantendo uma boa taxa de descida. Refazendo os experimentos com um numero maior de epochs, esses resultados ficam mais evidentes:

<img width="1055" height="253" alt="Image" src="https://github.com/user-attachments/assets/aca6fb4a-2f18-4e0b-94ec-880bfd3fef7a" />
